velox/functions/sparksql/String.h

```diff
@@ -455,4 +455,95 @@ struct SubstrFunction {
   }
 };
 
+template <typename T>
+struct OverlayFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& replace,
+      const int32_t pos,
+      const int32_t len) {
+    doCall<false>(result, input, replace, pos, len);
+  }
+
+  FOLLY_ALWAYS_INLINE void callAscii(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& replace,
+      const int32_t pos,
+      const int32_t len) {
+    doCall<true>(result, input, replace, pos, len);
+  }
+
+  template <bool isAscii>
+  FOLLY_ALWAYS_INLINE void doCall(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& replace,
+      const int32_t pos,
+      const int32_t len) {
+    // Calculate and append first part.
+    auto pair = substring<isAscii>(input, 1, pos - 1);
+    auto byteRange = stringCore::getByteRange<isAscii>(
+        input.data(), pair.first + 1, pair.second);
+    result.append(StringView(
+        input.data() + byteRange.first, byteRange.second - byteRange.first));
+
+    // Append second part.
+    result.append(replace);
+
+    // Calculate and append last part.
+    int32_t length = len >= 0 ? len : stringImpl::length<isAscii>(replace);
+    int64_t start = (int64_t)pos + (int64_t)length;
+    pair = substring<isAscii>(input, start, INT32_MAX);
+    byteRange = stringCore::getByteRange<isAscii>(
+        input.data(), pair.first + 1, pair.second);
+    result.append(StringView(
+        input.data() + byteRange.first, byteRange.second - byteRange.first));
+  }
+
+  // Information regarding the pos calculation:
+  // Hive and SQL use one-based indexing for SUBSTR arguments but also accept
+  // zero and negative indices for start positions. If a start index i is
+  // greater than 0, it refers to element i-1 in the sequence. If a start index
+  // i is less than 0, it refers to the -ith element before the end of the
+  // sequence. If a start index i is 0, it refers to the first element. Return
+  // pair of first indices and length.
+  template <bool isAscii>
+  FOLLY_ALWAYS_INLINE std::pair<int32_t, int32_t> substring(
+      const arg_type<Varchar>& input,
+      const int64_t pos,
+      const int64_t length) {
+    const int64_t len = stringImpl::length<isAscii>(input);
```

Comment:
I assume this code runs on VARBINARY inputs as well, in which case it is not valid to compute number of UTF-8 characters and we should use number of bytes. Please, check.

Reply:
Thanks, I got it.

When I try to define call function to distinguish Varchar and Varbinary like below, it will cause compiler error cause Varchar and Varbinary has same arg_type and out_type.
```
  FOLLY_ALWAYS_INLINE void call(
      out_type<Varchar>& result,
      const arg_type<Varchar>& input,
      const arg_type<Varchar>& replace,
      const int32_t pos,
      const int32_t len) {
    doCall<false, true>(result, input, replace, pos, len);
  }

  FOLLY_ALWAYS_INLINE void call(
      out_type<Varbinary>& result,
      const arg_type<Varbinary>& input,
      const arg_type<Varbinary>& replace,
      const int32_t pos,
      const int32_t len) {
    doCall<false, false>(result, input, replace, pos, len);
  }

template <bool isAscii, bool isVarchar>
  FOLLY_ALWAYS_INLINE void doCall(
      core::StringWriter& result,
      const StringView& input,
      const StringView& replace,
      const int32_t pos,
      const int32_t len) {}
```

Could you give me some suggestions or examples for how to define call functions when the function accept VARCHAR and VARBINARY at same time, like `func(VARCHAR, VARCHAR)` and `func(VARBINARY, VARBINARY)` ?

Reply:
@Yohahaha Hmm... you may have to write a Vector function. @laithsakka Is there a way to distinguish between VARCHAR and VARBINARY inputs in a simple function?

CC: @kagamiori

Reply:
here is example:
https://fburl.com/code/tzqb7wgu
ArraySubscriptSimpleFunc and ArraySubscriptSimpleFuncString are for the same function

```
 facebook::velox::registerFunction<
        ArraySubscriptSimpleFunc,
        double,
        Array<double>,
        int64_t>({"subscript_simple"});
        
    facebook::velox::registerFunction<
        ArraySubscriptSimpleFuncString,
        Varchar,
        Array<Varchar>,
        int64_t>({"subscript_simple"});
        ```

Reply:
Thanks for your suggestions!